### DISCUSSION TITLE
Multi node support in Torch-TRT

### DISCUSSION THREAD

**Discussion by apbose** [pytorch/TensorRT#3069] [RFCs]
Multi GPU compilation support
====================
## TL;DR

An important use case for Torch-TRT is to support multi-GPU and multi-node. The goal is to boost the performance using data parallelism and tensor parallelism to compile the model using multi gpu.  There are different ways to do this- fully sharded data parallelism, sequence parallelism and tensor parallelism. Data parallelism examples can be found in `/examples/distributed_inference/data_parallel_gpt2` and `/examples/distributed_inference/data_parallel_stable_diffusion`. This RFC focuses on tensor parallelism, an efficient model parallelism method for large model compilation.

## Goal

Accelerate the model compilation time using tensor parallelism

### Development of tensor parallel (TP) inference examples. 
The compilation of the model with tensor parallelism is agnostic to the franework as well as the model has been sharded properly by the network. The following two frameworks have been explored so far

- torch.distributed.tensor.parallel
Below is an example TP using `torch.distributed.tensor.parallel`
```markdown
https://github.com/pytorch/TensorRT/pull/3047/files#diffd70d4b88b03c3208178bf10992dfbb8e8ddb4ec6db247bbf42e0c2e706b0d5a5R5-R83
```
```python
from torch.distributed.tensor.parallel import ColwiseParallel, RowwiseParallel, parallelize_module
class ToyModel(torch.nn.Module()):
      ......
tp_model = ToyModel().to("cuda")
_world_size = int(os.environ["WORLD_SIZE"])
device_mesh = init_device_mesh(device_type="cuda", mesh_shape=(_world_size,))
_rank = device_mesh.get_rank()
tp_model = parallelize_module(
    module=tp_model,
    device_mesh=device_mesh,
    parallelize_plan={
        "in_proj": ColwiseParallel(input_layouts=Shard(0)),
        "out_proj": RowwiseParallel(output_layouts=Shard(0)),
        "in_proj2": ColwiseParallel(input_layouts=Shard(0)),
        "out_proj2": RowwiseParallel(output_layouts=Shard(0)),
    },
)
backend = "torch_tensorrt"
tp_model = torch.compile(
    tp_model,
    backend=backend,
    options={
        "truncate_long_and_double": True,
        "enabled_precisions": {torch.float32, torch.float16},
        "use_python_runtime": True,
        "min_block_size": 1
    },
    dynamic=False,
)
```
We start with this, primary reason being that `torch.distributed` is more stable and has no dependency on external libs. Due to this it does not introduce `torch.compile` graph breaks. The forward function is more in control. However the main limitation is that the sharding layout is manually set and will change from model to model.

We run the above using
`torchrun --nproc_per_node=2 tensor_parallel.py`

- Megatron LM
```python
from megatron import get_args, initialize_megatron, set_jit_fusion_options
from megatron.model import GPTModel
from megatron.training import train

# Initialize megatron
initialize_megatron(extra_args_provider=None, args_defaults={})
model = GPTModel(
        num_layers=args.num_layers,
        hidden_size=args.hidden_size,
        num_attention_heads=args.num_attention_heads,
        max_position_embeddings=args.max_position_embeddings
    )
backend = "torch_tensorrt"
tp_model = torch.compile(
    tp_model,
    backend=backend,
    options={
        "truncate_long_and_double": True,
        "enabled_precisions": {torch.float32, torch.float16},
        "use_python_runtime": True,
        "min_block_size": 1
    },
    dynamic=False,
)
```
Megatron-LM uses a JSON configuration file to set various parameters such as model size, sequence length, and parallelism settings.
This is a future path to be explored since it causes dynamic graph breaks in torch.compile. It however supports the mainstream LLM models via its configuration file. 

### Wrapping the NCCL ops in torch TensorRT converter library
NCCL  (NVIDIA Collective Communications Library) is a library designed to optimize collective communication operations across multiple GPUs and nodes. It is developed by NVIDIA and is widely used in distributed deep learning and high-performance computing to handle communication between GPUs efficiently. 

Since the NCCL collective communication is not supported in torch TRT, the above `torch.distributed` causes graph breaks, leading to slower compilation time as compared to torch in the first iteration while forming the TRT Engine. The following are the operations to be supported:
```markdown
https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/api/colls.html
```
Below are some examples through which the NCCL ops can be supported in torch TRT using NCCL based plugin. It uses TensorRT plugin registered under the namespace TRT_LLM_PLUGIN_NAMESPACE, which wraps the NCCL operations
```markdown
https://github.com/NVIDIA/TensorRT-LLM/blob/main/tensorrt_llm/functional.py#L3701-L4003
```

### Implemeting the above directly from TRT-LLM library

There can be two methods
1. Get the plugin code from tensorrt-llm and build the shared library from there
```python
try:
allgather_plg_creator = trt.get_plugin_registry().get_plugin_creator(
        "AllGather", "1", TRT_LLM_PLUGIN_NAMESPACE
 )
except Exception as e:
    _LOGGER.warning(
        "Unable to load .so"
    )
if(allgather_plg_creator is not None)
@dynamo_tensorrt_converter(<all_gather_nccl_target>)
 def aten_ops_nccl_send_plugin(
        ctx: ConversionContext,
        target: Target,
        args: Tuple[Argument, ...],
        kwargs: Dict[str, Argument],
        name: str,
    ) -> Union[TRTTensor, Sequence[TRTTensor]]:
        allgather_plg_creator = trt.get_plugin_registry().get_plugin_creator(
        "AllGather", "1", TRT_LLM_PLUGIN_NAMESPACE
        )

```

2. Install tensorrt-llm in torchTRT environment and load the plugin directly from there
```
def _load_plugin_lib():
    winmode = 0 if platform.system() == "Windows" else None
    handle = ctypes.CDLL(plugin_lib_path(),
                         mode=ctypes.RTLD_GLOBAL,
                         winmode=winmode)
    try:
        handle.initTrtLlmPlugins.argtypes = [ctypes.c_void_p, ctypes.c_char_p]
        handle.initTrtLlmPlugins.restype = ctypes.c_bool
    except AttributeError as err:
        raise ImportError('TensorRT-LLM Plugin is unavailable') from err
    assert handle.initTrtLlmPlugins(None,
                                    TRT_LLM_PLUGIN_NAMESPACE.encode('utf-8'))
```

**Comment by narendasan** [comment]
1. Add a utility that will download one of 3 wheels 

```py
tensorrt_llm-0.17.0.post1-cp310-cp310-win_amd64.whl
tensorrt_llm-0.17.0.post1-cp312-cp312-linux_aarch64.whl
tensorrt_llm-0.17.0.post1-cp312-cp312-linux_x86_64.whl
```
2. This utility is activated using the env var `USE_TRTLLM_PLUGINS=1`
3. If this is set then this function should be true provided we actually can DLOpen the .so https://github.com/pytorch/TensorRT/blob/5a4dd33ef261f76645d8bc585e4583a6624ffec8/py/torch_tensorrt/dynamo/conversion/custom_ops_converters.py#L23C1-L23C24

**Comment by narendasan** [comment]
Expose the use_aot_joint_export as an explicit setting for dynamo.compile / dynamo.convert_exported_program_to_serialized_trt_engine https://github.com/pytorch/TensorRT/blob/5a5d70726d3c81b6af230aa9a2aab57a272a89a5/py/torch_tensorrt/dynamo/_compiler.py#L415

**Comment by narendasan** [comment]
`support_dtensors` == `use_aot_joint_export=False` + `USE_TRTLLM_PLUGINS=1`

**Comment by narendasan** [comment]
Add unit tests to verify that at least compilation + the download utility work
